src/resources/parser/glb-container-resource.js

```diff
@@ -190,6 +192,45 @@ class GlbContainerResource {
         return GlbContainerResource.createSceneHierarchy(sceneClones, 'Entity');
     }
 
+    // get material variants
+    getMaterialVariants() {
+        return Object.keys(this.data.variants);
+    }
+
+    // apply material variant to entity
+    applyMaterialVariant(name, entity) {
+        const variant = this.data.variants[name];
+        if (variant === undefined) {
+            Debug.warn(`No variant named ${name} exists in resource`);
+            return;
+        }
+        const renders = entity.findComponents("render");
+        for (let i = 0; i < renders.length; i++) {
+            const renderComponent = renders[i];
+            this._applyMaterialVariant(variant, renderComponent.meshInstances);
+        }
+    }
+
+    // apply material variant to mesh instances
+    applyMaterialVariantInstances(name, instances) {
+        const variant = this.data.variants[name];
+        if (variant === undefined) {
+            Debug.warn(`No variant named ${name} exists in resource`);
+            return;
+        }
+        this._applyMaterialVariant(variant, instances);
+    }
+
+    // internally apply variant to instances
+    _applyMaterialVariant(variant, instances) {
+        instances.forEach((instance) => {
+            const meshVariants = this.data.meshVariants[instance.mesh.id];
+            if (meshVariants) {
+                instance.material = this.data.materials[meshVariants[variant]];
```

Comment:
please add `Debug.assert(instance.material);` just after this line, to validate it's not null, as those are errors that are harder to find.